src/software/ai/hl/stp/play/dynamic_plays/offense_play.cpp

```diff
@@ -0,0 +1,133 @@
+
+OffensePlay::OffensePlay(std::shared_ptr<Strategy> strategy)
+    : DynamicPlay(strategy),
+      attacker_tactic_(std::make_shared<AttackerTactic>(strategy)),
+      defense_play_(std::make_unique<DefensePlay>(strategy))
+{
+}
+
+void OffensePlay::terminate(const WorldPtr& world_ptr)
+{
+    DynamicPlay::terminate(world_ptr);
+    attacker_tactic_->terminate(world_ptr);
+}
+
+void OffensePlay::updateTactics(const PlayUpdate& play_update)
+{
+    PriorityTacticVector tactics;
+
+    // AttackerTactic should always be assigned
+    if (play_update.num_tactics > 0)
+    {
+        const bool attacker_not_suspended =
+            attacker_tactic_->tryResumingIfSuspended(play_update.world_ptr);
+
+        // The AttackerTactic should only be returned if it has not suspended
+        // execution of its current skill (it will not yield primitives if suspended)
+        if (attacker_not_suspended)
+        {
+            tactics.push_back({attacker_tactic_});
+        }
+
+        // Log visualize the state of the attacker's current skill
+        attacker_tactic_->visualizeSkillState(*play_update.world_ptr);
+    }
+
+    // Determine number of defense and support tactics to assign
+    auto [num_defenders, num_supporters] = assignNumOfDefendersAndSupporters(
+        std::max(static_cast<int>(play_update.num_tactics) - 1, 0));
+
+    // Get defense tactics from DefensePlay
+    std::vector<std::shared_ptr<Tactic>> defense_tactics;
+    defense_play_->updateTactics(PlayUpdate(
+        play_update.world_ptr, num_defenders,
+        [&](PriorityTacticVector new_tactics) {
+            for (auto& tactic_vec : new_tactics)
+            {
+                defense_tactics.insert(defense_tactics.end(), tactic_vec.begin(),
+                                       tactic_vec.end());
+            }
+        },
+        play_update.inter_play_communication,
+        play_update.set_inter_play_communication_fun));
+
+    // Create more support tactics if necessary
+    // (so that `support_tactics_` has `num_supporters` in it)
+    updateSupportTactics(num_supporters);
+
+    if (!support_tactics_.empty())
+    {
+        SkillState skill_state = attacker_tactic_->getSkillState();
+
+        std::vector<Point> existing_receiving_positions;
+        auto support_tactics_it = support_tactics_.begin();
+
+        // If the attacker has committed to a pass, set the receiving position
+        // of the first support tactic to the receiver point of the pass
+        if (skill_state.pass_committed && skill_state.pass)
+        {
+            Point receiving_position = skill_state.pass->receiverPoint();
+            support_tactics_.front()->updateReceivingPosition(receiving_position);
+            existing_receiving_positions.push_back(receiving_position);
+            ++support_tactics_it;
+        }
+
+        unsigned int num_positions_to_generate = static_cast<unsigned int>(
+            support_tactics_.size() - existing_receiving_positions.size());
```

Comment:
it seems like `existing_receiving_positions` is at most 1 right?

Reply:
Yes

Reply:
We are using `existing_receiving_positions` similarly. iirc in `ShootOrPassPlay` the receivers are actually `MoveTactics` and there is only 1 `ReceiverTactic` that is always assigned to the current best pass. Hence you need to put the `ReceiverTactic` position in `existing_receiving_positions` so that the `MoveTactics` don't go to the same area as the `ReceiverTactic`.

In the new `OffensePlay`, all receivers are `ReceiverTactic`. So if the attacker has committed to a pass, we assign 1 `ReceiverTactic` to the pass receiver point and put that receiver point in `existing_receiving_positions` so that other receivers don't go to the same area.

Reply:
if it's at most one, why not make it an optional

Reply:
Oh I see why, it's because of the `ReceiverPassGenerator`